### PR TITLE
a post

### PR DESCRIPTION
**Pull request by alaiacano**
Ask for reviewers:

The writing is in `_posts/2021-08-15-interviews.markdown`. Feel free to leave as much feedback as you want.

It's hard to see the code interwoven into the text because the gists don't expand. You could save the [raw html](https://raw.githubusercontent.com/alaiacano/alaiacano.github.io/linkedlist/_site/interviews/2021/08/10/interviews.html) locally and open that file in a browser if you want. Or maybe something else?

The code in `code_examples/linked_list` should work! If you want to give normal code review on it I'd be forever grateful (I'll update the gists that get included into the post itself).

Ignore all the other config stuff.

Thanks!

### CODE REVIEW

File: _posts/2021-08-15-interviews.markdown [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+---
+
+For most of us, interviewing is hard. And annoying, frustrating, and stressful. Most interviews these days consist of a string of 1-hour interviews covering:
+
+- Programming (sometimes 2 of these).
+- Systems design.
+- 1-2 specialization-specific interviews: ML, Data Engineering, Frontend, Distributed Systems, etc.
+- An interview with a hiring manager. Sometimes this is about "leadership" or other general topics depending on the role.
+
+The programming interview is probably the most hated. I personally don't like being on either side of these interviews. Leetcode-style problems that test _CS fundamentals_ can be fun puzzles for some people, and are arguably a decent evaluation of junior engineers without industry experience, but it's a common complaint among senior engineers that they have to study material that they haven't used in years to get a job where they won't use it. This tweet, from the author of [Homebrew](https://brew.sh/), lives in infamy.
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown]: > a common complaint among senior engineers that they have to study material that they haven't used in years to get a job where they won't use it

I wonder if you have any personal anecdotes that could be interesting to mention here? like a time when you were asked in an interview to prove you knew algorithm X in a role that would never use it

File: _posts/2021-08-15-interviews.markdown [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+<blockquote class="twitter-tweet" data-dnt="true" data-theme="dark"><p lang="en" dir="ltr">Google: 90% of our engineers use the software you wrote (Homebrew), but you can’t invert a binary tree on a whiteboard so fuck off.</p>&mdash; Max Howell (@mxcl) <a href="https://twitter.com/mxcl/status/608682016205344768?ref_src=twsrc%5Etfw">June 10, 2015</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+There are a few questions that people like to use as punching bags. Invert a binary tree. Reverse a linked list. Perform breadth-first and/or depth-first tree traversals. Anything to do with heaps. I recently read the [history of the "Reverse A Linked List" problem](https://www.hillelwayne.com/post/linked-lists/) and find this take fascintating. The goal wasn't to reverse a linked list, it was to make sure you are familiar with pointer manipulation in c, and reversing the linked list was a good way to demonstrate that.
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown]:
```suggestion
There are a few questions that people like to use as punching bags. Invert a binary tree. Reverse a linked list. Perform breadth-first and/or depth-first tree traversals. Anything to do with heaps. I recently read the [history of the "Reverse A Linked List" problem](https://www.hillelwayne.com/post/linked-lists/) and find this take fascintating. The goal wasn't to reverse a linked list, it was to make sure you are familiar with pointer manipulation in C, and reversing the linked list was a good way to demonstrate that.
```

File: _posts/2021-08-15-interviews.markdown [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+## Reversing a Linked List
+
+We'll start with the classic algorithm question. My solution is [here](https://gist.github.com/alaiacano/35a63e7631b60641fc0342f31461b80d#file-ll-py). Inserting new nodes at the head of the list is `O(1)` and reversing the list is `O(N)`, and does so by moving pointers around and not making a full copy of the list in memory.
+
+To populate, print, reverse, and then print the list again, we'd do this:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=run_ll.py" type="text/javascript"></script>
+
+Wonderful. In the real world, we would have used [`List.reverse`](https://github.com/python/cpython/blob/e5c8ddb1714fb51ab1defa24352c98e0f01205dc/Objects/listobject.c#L1042-L1056) from the standard library, but this produces the same result. Would we have run the code in a module like this? Probably not!
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown:33]: > Would we have run the code in a module like this? Probably not!

do you mean that the code in `run_ll.py` doesn't seem like something you'd need to write in real life? I had trouble understanding this part

File: _posts/2021-08-15-interviews.markdown [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+## Enter the LinkedList Pipeline
+
+Let's rewrite how we use this `LinkedList` class into something a little more _modern_:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=pipeline1.yaml" type="text/javascript"></script>
+
+YAML! Nice. Yaml is fun to joke about, but Vicki Boykis posted [a twitter poll](https://twitter.com/vboykis/status/1407821631997759488) about which languages people use in Data Science/ML and the most common reply was "does yaml count?" If everyone asks, it means yes.
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown]: you format `yaml` 3 different ways in this line 😀

File: _posts/2021-08-15-interviews.markdown [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+The above yaml does not actually do anything, and this is ostensibly a post about software coding interviews, so let's build something that reverses a linked list given this pipeline definition. Is this overkill for the goal of reversing a 4-element linked list? Sure. But it's a good analog to the modular systems that we are paid to build and use as software engineers (especially Data/ML engineers). It's the kind of thing that we actually do every day, and should be familiar with. In other words, **Components are the new Pointers.**
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown]: I think this point could use a little more elaboration, something like: 

its rare in an actual job in 2021 to write functions like "reverse a linked list" which are provided in standard libraries 99.9% of the time, but what is more common is to assemble various pre-made components together into a larger whole to make something out of them, like this example "pipeline-as-yaml". So lets use this toy example of "reverse a linked list" to demonstrate an interview problem that would have you write a framework for assembling these components into a larger whole (a "pipeline") - since _this_ is likely to more closely resemble the day-to-day work of the job the interview is for.

can also say something like - ok so if this YAML is the interface that LinkedList users will use in our framework, how do we write the code that implements this framework?

File: _posts/2021-08-15-interviews.markdown [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+So let's dive in.
+
+### Coding problem: Implement a framework to execute this pipeline
+
+First we'll need to define a `Task` class to carry out the logic of each `task` defined in the yaml. A `Task` may take an input `LinkedList` from a previous `Task`, performs some operation on it, and produces an output. Here's the interface:
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=base_task.py" type="text/javascript"></script>
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown:51]: might be worth adding a brief comment or two in the code here describing what `abc` and `@abstractmethod` are doing there, I always feel slightly confused by `abc` since I don't write enough Python and have to re-remember what the point of it is from context clues

File: _posts/2021-08-15-interviews.markdown [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+It takes an input `LinkedList` and a `name` for the task, and has an abstract `execute` method that will have to mutate `self._lst` (which is maybe not ideal but we'll work with it). Let's extend `Task` to perform the actions seen in the yaml: `push_values`, `reverse_list`, and `print_list`.
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=tasks.py" type="text/javascript"></script>
+
+You can see that I moved the `reverse` method into `ReverseListTask`, so we can remove the method from the `LinkedList` class. One benefit of this task-based approach is that the logic that we want to perform is located within the `Task`, and this helps us achieve that task. None of the other `Task`s need to know about reversing a list, and having code around to do unnecessary tasks might become step towards "Dependency Hell."
+
+Our yaml config has an `action` field that we can use to decide which of these subclasses to use for each step in the pipeline. Let's use a factory function for that <sup id="a1">[1](#f1)</sup>.
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=task_factory.py" type="text/javascript"></script>
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown:63]: I was going to post a comment here making a suggestion on the interface you have in your Task class, which was: would it be easier if instead of each Task having a reference to `lst`, you reworked the interface so the `execute` method received the `lst` to operate on? Then you don't have to pass `lst` to the TaskFactory so it can pass it to each task ...

but this made me think of a broader point you could perhaps work into the post - that by giving the candidate a more concrete task where they have much more of an opportunity to express themselves and be creative, it brings up the chance to discuss interesting aspects of the code they wrote that you don't get when they are just mechanically implementing a algorithm - like, why did you design the interface so each `Task` has a reference to `lst` in the constructor, versus this other idea? what are the trade-offs of each? etc. 

in other words, it'd be interesting to hammer a bunch in the post on why this approach is better - its more faithful to the actual work the candidate will do, it creates plenty of chances for discussion points, etc

File: _posts/2021-08-15-interviews.markdown [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=pipeline2.yaml" type="text/javascript"></script>
+
+Now we can implement a `Parser2` that's a little more advanced. Rather than a `List[Task]` we'll create a dictionary from `task_id -> List[downstream Tasks]`, and step through all of the tasks. Here's how `run` is implemented now (see the link at the bottom for the full code):
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=parser2.py" type="text/javascript"></script>
+
+I made a _graph_ of parent and child tasks, and then step through starting at the head and hitting every child, passing the output of the parent to each of its children tasks. **It's a depth-first graph traversal** just hiding there right in front of us!
+
+## Wrap-up
```
Comment (mattnworb) [inline _posts/2021-08-15-interviews.markdown:153]: as mentioned above I think you could reiterate here the various reasons why you think this approach is better, such as

- the task is more life-like
- lots of interesting stuff to talk about in how they design the code, interface, etc
- lots of ways to evolve the question to do other interesting things they'd also have to tackle in the real world some day, like the graph stuff

**Review by mattnworb — COMMENTED** [review]
Jekkyl looks like fun 😂 

I think this is looking like an interesting post! I have a few small suggestions and comments below.
